internal/btf/ext_info.go

```diff
@@ -310,44 +273,85 @@ func parseFuncInfoRecords(r io.Reader, bo binary.ByteOrder, recordSize uint32, r
 			return nil, fmt.Errorf("offset %v is not aligned with instruction size", fi.InsnOff)
 		}
 
+		// ELF tracks offset in bytes, the kernel expects raw BPF instructions.
+		// Convert as early as possible.
+		fi.InsnOff /= asm.InstructionSize
+
 		out = append(out, fi)
 	}
 
 	return out, nil
 }
 
+var LineInfoSize = uint32(binary.Size(bpfLineInfo{}))
+
 // LineInfo represents the location and contents of a single line of source
 // code a BPF ELF was compiled from.
 type LineInfo struct {
-	// Instruction offset of the function within an ELF section.
-	// After parsing a LineInfo from an ELF, this offset is relative to
-	// the function body instead of an ELF section.
+	fileName   string
+	line       string
+	lineNumber uint32
+	lineColumn uint32
+
+	// TODO: We should get rid of the fields below, but for that we need to be
+	// able to write BTF.
+
+	// Instruction offset of the line within its enclosing function, in instructions.
+	insnOff     uint32
+	fileNameOff uint32
+	lineOff     uint32
+}
+
+type bpfLineInfo struct {
+	// Instruction offset of the line within the whole instruction stream, in instructions.
 	InsnOff     uint32
 	FileNameOff uint32
 	LineOff     uint32
 	LineCol     uint32
 }
 
+func (li *LineInfo) FileName() string {
+	return li.fileName
+}
+
+func (li *LineInfo) Line() string {
+	return li.line
+}
+
+func (li *LineInfo) LineNumber() uint32 {
+	return li.lineNumber
+}
+
+func (li *LineInfo) LineColumn() uint32 {
+	return li.lineColumn
+}
+
+func (li *LineInfo) String() string {
+	return li.line
+}
+
 // Marshal writes the binary representation of the LineInfo to w.
 // The instruction offset is converted from bytes to instructions.
-func (li LineInfo) Marshal(w io.Writer, offset uint64) error {
-	li.InsnOff += uint32(offset)
-	// The kernel expects offsets in number of raw bpf instructions,
-	// while the ELF tracks it in bytes.
-	li.InsnOff /= asm.InstructionSize
-	return binary.Write(w, internal.NativeEndian, li)
+func (li *LineInfo) Marshal(w io.Writer, offset uint64) error {
+	// TODO: Error out if lineNumber, lineColumn are out of bounds.
```

Comment:
Also TODO

Reply:
I don't think you have to worry about out-of-bound `lineNumber`s and `lineColumn`s, they are never used as array/slice index and verifying them would require you to inspect the original source code which might very well not be available on the machine loaded the ELF file.

Reply:
True, I was thinking more along the lines of robustness when we let people generate their own context.